components/mpas-ocean/src/shared/mpas_ocn_vertical_remap.F

```diff
@@ -0,0 +1,436 @@
+      ! Currently, zstar coordinate is hard-coded
+      call ocn_vert_regrid(restingThickness, layerThickness, &
+                               layerThicknessNew, err)
+
+      ! There's currently a silly issue in PPR: it assumes the "coordinates" for
+      ! the data being remapped increases with index. Here, we have the opposite
+      ! (z becomes more -ve with k)...
+      ! I'll make some updates to PPR to handle this, but for now, just flip the
+      ! signs to make PPR happy. This shouldn't have any actual impact.
+
+      !$omp parallel
+      !$omp do schedule(runtime) private(k)
+      do iCell = 1, nCells
+         heightCellNow(:, iCell) = bottomDepth(iCell)
+         heightCellNew(:, iCell) = bottomDepth(iCell)
+
+         ! reconstruct "now" heights
+         do k = maxLevelCell(iCell), minLevelCell(iCell), -1
+            heightCellNow(k, iCell) = &
+               heightCellNow(k + 1, iCell) - layerThickness(k, iCell)
+         end do
+
+         ! reconstruct "new" heights
+         heightCellNew(minLevelCell(iCell), iCell) = heightCellNow(minLevelCell(iCell), iCell)
+         do k = maxLevelCell(iCell), minLevelCell(iCell)+1, -1
+            heightCellNew(k, iCell) = &
+               heightCellNew(k + 1, iCell) - layerThicknessNew(k, iCell)
+         end do
+      end do
+      !$omp end do
+      !$omp end parallel
+
+      call ocn_diagnostic_solve_layerThicknessEdge(normalVelocity, layerThickness, layerThickEdge)
+      call ocn_diagnostic_solve_layerThicknessEdge(normalVelocity, layerThicknessNew, layerThickEdgeNew)
+
+      !$omp parallel
+      !$omp do schedule(runtime) private(k,kTop,kBot,cell1,cell2)
+      do iEdge = 1, nEdges
+
+         cell1 = cellsOnEdge(1,iEdge)
+         cell2 = cellsOnEdge(2,iEdge)
+
+         kTop = minLevelEdgeBot(iEdge)
+         kBot = maxLevelEdgeTop(iEdge)
+
+         ! We only need to define the kBot+1 depth for one active cell
+         ! as the absolute kBot+1 depth has no effect on remapping
+         if (cell1 <= nCellsAll .and. maxLevelCell(cell1) > 0) then
+            heightEdgeNow(:, iEdge) = heightCellNow(kBot+1, cell1)
+            heightEdgeNew(:, iEdge) = heightCellNew(kBot+1, cell1)
+         elseif (cell2 <= nCellsAll .and. maxLevelCell(cell2) > 0) then
+            heightEdgeNow(:, iEdge) = heightCellNow(kBot+1, cell2)
+            heightEdgeNew(:, iEdge) = heightCellNew(kBot+1, cell2)
+         else
+            call mpas_log_write('No valid cell for edge remapping', MPAS_LOG_WARN)
+            heightEdgeNow(:,iEdge) = 0.0_RKIND
+            cycle
+         endif
+
+         do k = kBot, kTop, -1
+            heightEdgeNow(k, iEdge) = heightEdgeNow(k+1, iEdge) - layerThickEdge(k, iEdge)
+            heightEdgeNew(k, iEdge) = heightEdgeNew(k+1, iEdge) - layerThickEdgeNew(k, iEdge)
+         end do
+
+      end do
+      !$omp end do
+      !$omp end parallel
+
+      ! Assign new layerThicknesses
+      layerThickness = layerThicknessNew
```

Comment:
I'm seeing complaints about this line in debug mode.
```
Error termination. Backtrace:
At line 213 of file mpas_ocn_vertical_remap.F
Fortran runtime error: Array bound mismatch for dimension 2 of array 'layerthickness' (873/872)

Error termination. Backtrace:
#0  0x55fe5e03e21f in __ocn_vertical_remap_MOD_ocn_remap_vert_state
	at /home/xylar/code/compass/compass/ocn/add-vertical-remap-feature/components/mpas-ocean/src/shared/mpas_ocn_vertical_remap.F:213
```
I think the problem is that layerThickness has size `nCellsAll+1` to accommodate a dummy cell, whereas `layerThicknessNew` has been allocated with size `nCellsAll`.  One way to handle this would be:
```suggestion
      layerThickness(:, 1:nCells) = layerThicknessNew
```

Reply:
Hmm. Naive question: How does `layerThickness` come to have dimension of `nCellsAll+1` when the registry says its dimensions are `nVertLevels` by `nCells`?

Reply:
I know, this is super frustrating.  I searched the code and I can't figure out where this actually happens but there is a dummy cell, edge and vertex for every field with `nCells`, `nEdges` or `nVertices` as a dimension.  The idea was that this cell (or edge or vertex) would be given a safe dummy value such that it could be supplied as the neighbor in situations where there is no actual neighbor.  Again, the idea was that it would be faster to compute on a dummy neighbor than to have checks to find out if a given neighbor existed or not.  I don't know how practical that approach has ended up being but its legacy remains both in the size of the arrays and in the need to check whether neighbors are `<= nCellsAll`, since an invalid (i.e. dummy) neighbor is given index `nCellsAll+1`.